File: 0112-path-sum/0112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool helper(TreeNode* root, int& sum, int targetSum){
        if (root == nullptr){
            return false;
        }
        sum += root->val;
        cout << "sum up: " <<sum<<"\n";
        
        if (root->left == nullptr && root->right == nullptr && sum == targetSum){
            return true;  
        }

        if (helper(root->left, sum, targetSum) || helper(root->right, sum, targetSum)){
            return true;
        }
        
        sum -= root->val;
        cout << "sum down :" <<sum << endl;
        return false;
    }
    bool hasPathSum(TreeNode* root, int targetSum) {
        int sum = 0;
        return helper(root, sum, targetSum);
    }
};

File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self, root, current_sum, target_sum):
        if root is None:
            return False
        
        current_sum += root.val
        
        if root.left is None and root.right is None and current_sum == target_sum:
            return True
        
        if self.helper(root.left, current_sum, target_sum) or self.helper(root.right, current_sum, target_sum):
            return True
        
        current_sum -= root.val
        
        return False
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        return self.helper(root, 0, targetSum)